src/Catel.MVVM/Collections/ObservableDictionary.cs

```diff
@@ -0,0 +1,682 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="ObservableDictionary.cs" company="Catel development team">
+//   Copyright (c) 2008 - 2019 Catel development team. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Catel.Collections
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Collections.ObjectModel;
+    using System.Collections.Specialized;
+    using System.ComponentModel;
+    using System.Linq;
+    using System.Runtime.Serialization;
+    using IoC;
+    using Logging;
+    using Services;
+
+    internal class IndexedKeyedCollection<TKey, TValue> : KeyedCollection<TKey, KeyValuePair<TKey, TValue>>
+    {
+        public IndexedKeyedCollection()
+        {
+        }
+
+        public IndexedKeyedCollection(IEqualityComparer<TKey> comparer)
+            : base(comparer)
+        {
+        }
+
+        protected override TKey GetKeyForItem(KeyValuePair<TKey, TValue> item)
+        {
+            return item.Key;
+        }
+    }
+
+    /// <summary>
+    /// Represents an observable collection of keys and values.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+#if NET || NETCORE
+    [Serializable]
+#endif
+    public class ObservableDictionary<TKey, TValue> : IDictionary<TKey, TValue>,
+        IDictionary,
+        IReadOnlyDictionary<TKey, TValue>,
+        ISerializable,
+        IDeserializationCallback,
+        INotifyCollectionChanged,
+        INotifyPropertyChanged
+    {
+        private static readonly ILog Log = LogManager.GetCurrentClassLogger();
+
+        private readonly IndexedKeyedCollection<TKey, TValue> _collection;
+
+        private readonly Lazy<IDispatcherService> _dispatcherService = new Lazy<IDispatcherService>(() => IoCConfiguration.DefaultDependencyResolver.Resolve<IDispatcherService>());
```

Comment:
Do we use the same tech in FastObservableCollection? Preferrably we don't resolve anything in a static manner, but sometimes we don't have a choice. But at least we can keep stuff similar in different classes.

Reply:
As the dispatcher is used in a few places and dependency injection construction is not the best course of action for this class, this was the best pathway to avoid calling the resolver multiple times. The same behavior is also implemented in [FastObservableCollection.cs:L33](https://github.com/Catel/Catel/blob/develop/src/Catel.MVVM/Collections/FastObservableCollection.cs#L33).